sdk/samples/core/rpc_client_pending_commands.h

```diff
@@ -0,0 +1,175 @@
+  }
+  else
+  {
+    // NOLINTNEXTLINE(bugprone-misplaced-widening-cast)
+    *out_clock_msec = (int64_t)((time(NULL)) * 1000);
+  }
+#endif
+  return AZ_OK;
+}
+
+AZ_INLINE az_result add_command(
+    pending_commands_array* pending_commands,
+    az_span correlation_id,
+    az_span command_name,
+    int32_t timeout_ms)
+{
+  az_result ret = AZ_OK;
+  if (pending_commands->pending_commands_count >= RPC_CLIENT_MAX_PENDING_COMMANDS)
+  {
+    printf(LOG_APP_ERROR "Pending Commands Array already has max number of commands.\n");
+    return AZ_ERROR_OUT_OF_MEMORY;
+  }
+  int32_t empty_index = -1;
+  for (int i = 0; i < RPC_CLIENT_MAX_PENDING_COMMANDS; i++)
+  {
+    if (az_span_size(pending_commands->commands[i].command_name) == 0)
+    {
+      empty_index = i;
+      break;
+    }
+  }
+  if (empty_index < 0)
+  {
+    return AZ_ERROR_OUT_OF_MEMORY;
+  }
+
+  pending_commands->pending_commands_count++;
+  az_span_copy(pending_commands->commands[empty_index].correlation_id, correlation_id);
+  pending_commands->commands[empty_index].command_name = command_name;
+
+  int64_t clock = 0;
+  ret = clock_msec(&clock);
+  pending_commands->commands[empty_index].context
+      = az_context_create_with_expiration(&az_context_application, clock + timeout_ms);
+
+  return ret;
+}
+
+AZ_INLINE az_result remove_command(pending_commands_array* pending_commands, az_span correlation_id)
+{
+  for (int i = 0; i < RPC_CLIENT_MAX_PENDING_COMMANDS; i++)
+  {
+    if (az_span_is_content_equal(pending_commands->commands[i].correlation_id, correlation_id))
+    {
+      az_context_cancel(&pending_commands->commands[i].context);
+      az_span_fill(pending_commands->commands[i].correlation_id, 0x0);
+      pending_commands->commands[i].command_name = AZ_SPAN_EMPTY;
+      pending_commands->pending_commands_count--;
+      return AZ_OK;
+    }
+  }
+  return AZ_ERROR_ITEM_NOT_FOUND;
+}
+
+AZ_INLINE bool is_pending_command(pending_commands_array pending_commands, az_span correlation_id)
+{
+  for (int i = 0; i < RPC_CLIENT_MAX_PENDING_COMMANDS; i++)
+  {
+    if (az_span_is_content_equal(pending_commands.commands[i].correlation_id, correlation_id))
+    {
+      return true;
+    }
+  }
+  return false;
+}
+
+AZ_INLINE pending_command* get_first_expired_command(pending_commands_array pending_commands)
+{
+  pending_command* expired_command = NULL;
+  int64_t current_time = 0;
+  if (az_result_failed(clock_msec(&current_time)))
+  {
+    return NULL;
+  }
+  for (int i = 0; i < RPC_CLIENT_MAX_PENDING_COMMANDS; i++)
+  {
+    if (az_span_size(pending_commands.commands[i].command_name) > 0
+        && az_context_has_expired(&pending_commands.commands[i].context, current_time))
+    {
+      if (expired_command == NULL
+          || az_context_get_expiration(&pending_commands.commands[i].context)
+              < az_context_get_expiration(&expired_command->context))
+      {
+        expired_command = &pending_commands.commands[i];
```

Comment:
break; ?

Reply:
not desired because we want to get the most expired command so we expire them in order and never have an old one hang around a weirdly long amount of time because it keeps not getting returned first in this function

Reply:
please add comments stated above